### PR TITLE
add DI container support

### LINKED ISSUE #324: Is it possible to use the DI Container inside a task?
With that I wouldn't need to manually create service objects inside the task 😄

### PR DESCRIPTION
**Pull request by LordSimal**
Fixes #324 

This implementation allows the user to get access to the DI container instance via adding

```
public function services(ContainerInterface $container): void {
	$this->testService = $container->get(TestService::class);
}
```

to the desired task.

I first thought to implement the same behaviour as it is present in the controller where the `services()` method automatically detects the desired services via the argument types but i think it is easier to just pass the DI container instance to the task and the user can retrieve as many services as he wants from that instance.

The only thing I am not really sure if its the "correct" way is the fact how I inject the container instance into the processor (and therefore into the task) - see bellow my github code comment.

### CODE REVIEW

File: src/Command/RunCommand.php [hunk 1]
```diff
@@ -17,6 +18,19 @@
  */
 class RunCommand extends Command {
 
+	/**
+	 * @var \Cake\Core\ContainerInterface
+	 */
+	public $container;
```
Comment (dereuromark) [inline src/Command/RunCommand.php:24]: Does this need to be public?

File: tests/test_app/src/Queue/Task/FooTask.php [hunk 2]
```diff
@@ -20,6 +23,11 @@ class FooTask extends Task {
 	 */
 	public $retries = 1;
 
+	/**
+	 * @var \TestApp\Services\TestService
+	 */
+	public $testService;
```
Comment (dereuromark) [inline tests/test_app/src/Queue/Task/FooTask.php:29]: Same here

File: src/Command/RunCommand.php [hunk 3]
```diff
@@ -21,7 +21,7 @@ class RunCommand extends Command {
 	/**
 	 * @var \Cake\Core\ContainerInterface
 	 */
-	public $container;
+	private $container;
```
Comment (dereuromark) [inline src/Command/RunCommand.php:24]: never private ;) always protected

File: src/Command/RunCommand.php [hunk 4]
```diff
@@ -17,6 +18,19 @@
...
 
+	/**
+	 * @var \Cake\Core\ContainerInterface
+	 */
+	protected $container;
+
+	/**
+	 * @param \Cake\Core\ContainerInterface $container
+	 */
+	public function __construct(ContainerInterface $container){
```
Comment (dereuromark) [inline src/Command/RunCommand.php:29]:
```suggestion
	public function __construct(ContainerInterface $container) {
```
I am still confused why CS is not working (anymore)

File: src/Queue/TaskInterface.php [hunk 5]
```diff
@@ -22,4 +24,12 @@ interface TaskInterface {
 	 */
 	public function run(array $data, int $jobId): void;
 
+	/**
+	 * Provides access to the DI container and its registered services
+	 *
+	 * @param \Cake\Core\ContainerInterface $container The DI container instance
+	 * @return void
+	 */
+	public function services(ContainerInterface $container): void;
```
Comment (dereuromark) [inline src/Queue/TaskInterface.php:33]: This sounds like we need to make it a major... 
I wonder if there is sth we could do before we major (for Cake5)

Comment (dereuromark) [reply src/Queue/TaskInterface.php:33]: Maybe we could just let users add a Trait (e.g. ServicesTrait) that could provide this as opt-in?

File: tests/test_app/src/Queue/Task/FooTask.php [hunk 6]
```diff
@@ -30,7 +41,23 @@ class FooTask extends Task {
...
-		$this->out('CakePHP Foo Example.');
+		$this->io->out('CakePHP Foo Example.');
+		$test = $this->testService->output();
+		$this->io->out($test);
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function services(ContainerInterface $container): void {
```
Comment (dereuromark) [inline tests/test_app/src/Queue/Task/FooTask.php:52]: Why is this needed?
Wouldnt it be easier if the trait directly exposes the container method to get sth you need?

File: src/Queue/ServicesTrait.php [hunk 7]
```diff
@@ -7,12 +7,26 @@
...
+	 */
+	protected $container;
+
+	/**
+	 * @param string $id Classname or identifier of the service you want to retrieve
+	 * @return mixed
+	 * @throws \Psr\Container\NotFoundExceptionInterface
+	 * @throws  \Psr\Container\ContainerExceptionInterface
+	 */
+	public function getService(string $id) {
```
Comment (dereuromark) [inline src/Queue/ServicesTrait.php:20]: Why do those need public access I wonder

**Comment by dereuromark** [comment]
Sounds pretty BC safe to me now.

**Comment by LordSimal** [comment]
Should I ignore that phpcs error regarding `Invalid Inline Doc Block`? Or what exactly is wrong there?

File: src/Queue/Processor.php [hunk 8]
```diff
@@ -186,6 +194,11 @@ protected function runJob(QueuedJob $queuedJob, string $pid): void {
 
 			$data = $queuedJob->data ? unserialize($queuedJob->data) : null;
 			$task = $this->loadTask($taskName);
+			$traits = class_uses($task);
+			if ($this->container && $traits && in_array(ServicesTrait::class, $traits, true)) {
+				/** @phpstan-ignore-next-line */
```
Comment (dereuromark) [inline src/Queue/Processor.php:199]: Why is this needed?

Comment (dereuromark) [reply src/Queue/Processor.php:199]: We ignore it, all good

**Comment by dereuromark** [comment]
Can you maybe add some docs for it too?
So people know how to use it?
